pallets/avn-anchor/src/lib.rs

```diff
@@ -288,9 +319,47 @@ pub mod pallet {
 
             Ok(())
         }
+
+        #[pallet::weight(<T as pallet::Config>::WeightInfo::set_checkpoint_fee())]
+        #[pallet::call_index(6)]
+        pub fn set_checkpoint_fee(
+            origin: OriginFor<T>,
+            chain_id: ChainId,
+            new_fee: BalanceOf<T>,
+        ) -> DispatchResult {
+            ensure_root(origin)?;
+
+            CheckpointFee::<T>::insert(chain_id, new_fee);
+            Self::deposit_event(Event::CheckpointFeeUpdated { chain_id, new_fee });
+
+            Ok(())
+        }
     }
 
     impl<T: Config> Pallet<T> {
+        pub(crate) fn charge_fee(handler: T::AccountId, chain_id: ChainId) -> DispatchResult {
+            let checkpoint_fee = Self::get_checkpoint_fee(chain_id);
+
+            T::FeeHandler::pay_treasury(&checkpoint_fee, &handler)?;
+
+            Self::deposit_event(Event::CheckpointFeeCharged {
+                handler: handler.clone(),
+                fee: checkpoint_fee,
+                chain_id,
+            });
+
+            Ok(())
+        }
+
+        pub fn get_checkpoint_fee(chain_id: ChainId) -> BalanceOf<T> {
+            let fee = CheckpointFee::<T>::get(chain_id);
+            if CheckpointFee::<T>::contains_key(chain_id) {
+                fee
+            } else {
+                T::DefaultCheckpointFee::get()
```

Comment:
I wonder if we could make the CheckpointFee a ValueQuery with T::DefaultCheckpointFee::get() being the default value, to simplify this and eliminate the helper function.

Reply:
will try it out, thanks